File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
from collections import deque

m, n, k = map(int, input().split())     # 세로, 가로, 영역 개수
arr = [[0]*n for _ in range(m)]
for _ in range(k):
    sx, sy, ex, ey = map(int, input().split())
    for i in range(sy, ey):
        for j in range(sx, ex):
            arr[i][j] = 1

def bfs(y, x):
    q = deque()
    q.append((y, x))
    arr[y][x] = 1
    
    size = 1
    dy = [0, 0, -1, 1]
    dx = [-1, 1, 0, 0]
    while q:
        y, x = q.popleft()
        for i in range(4):
            ny = y+dy[i]
            nx = x+dx[i]
            if 0<=ny<m and 0<=nx<n and arr[ny][nx]==0:
                arr[ny][nx] = 1
                q.append((ny, nx))
                size += 1
    result.append(size)

result = []
for i in range(m):
    for j in range(n):
        if arr[i][j]==0:
            bfs(i, j)

result.sort()
print(len(result))
for i in result:
    print(i, end=' ')

File: 백준/Silver/2583. 영역 구하기/영역 구하기.java
import java.util.*;
import java.io.*;

public class Main {

    static int n, m;
    static int[][] arr;
    static boolean[][] visited;
    static ArrayList<Integer> answer = new ArrayList<>();
    static class Pair {
        int y, x;

        public Pair(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }
    static int[] dy = {0, 1, 0, -1};
    static int[] dx = {1, 0, -1, 0};

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        arr = new int[n][m];
        visited = new boolean[n][m];
        while (k-->0) {
            st = new StringTokenizer(br.readLine());
            int sx = Integer.parseInt(st.nextToken());
            int sy = Integer.parseInt(st.nextToken());
            int ex = Integer.parseInt(st.nextToken());
            int ey = Integer.parseInt(st.nextToken());
            for (int i = sy; i<ey; i++) {
                for (int j = sx; j<ex; j++) {
                    arr[i][j] = 1;
                }
            }
        }
        for (int i = 0; i<n; i++) {
            for (int j = 0; j<m; j++) {
                if (arr[i][j]==0 && !visited[i][j]) {
                    answer.add(bfs(i, j));
                }
            }
        }

        Collections.sort(answer);
        System.out.println(answer.size());
        for (int i = 0; i<answer.size(); i++) {
            System.out.print(answer.get(i) + " ");
        }
    }

    public static int bfs(int y, int x) {
        int count = 1;
        visited[y][x] = true;
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(y, x));

        while (!q.isEmpty()) {
            Pair curr = q.poll();
            for (int d = 0; d<4; d++) {
                int ny = curr.y+dy[d];
                int nx = curr.x+dx[d];
                if (ny<0 || ny>=n || nx<0 || nx>=m) continue;
                if (!visited[ny][nx] && arr[ny][nx]==0) {
                    q.add(new Pair(ny, nx));
                    count++;
                    visited[ny][nx] = true;
                }
            }
        }

        return count;
    }
}